src/app/clusters/user-label-server/user-label-server.cpp

```diff
@@ -57,17 +58,24 @@ UserLabelAttrAccess gAttrAccess;
 CHIP_ERROR UserLabelAttrAccess::ReadLabelList(EndpointId endpoint, AttributeValueEncoder & aEncoder)
 {
     CHIP_ERROR err = CHIP_NO_ERROR;
-    DeviceLayer::AttributeList<app::Clusters::UserLabel::Structs::LabelStruct::Type, DeviceLayer::kMaxUserLabels> labelList;
 
-    if (DeviceLayer::PlatformMgr().GetUserLabelList(endpoint, labelList) == CHIP_NO_ERROR)
+    DeviceLayer::DeviceInfoProvider::UserLabelIterator * it = DeviceLayer::GetDeviceInfoProvider()->IterateUserLabel(endpoint);
+
+    if (it && it->Count() > 0)
     {
-        err = aEncoder.EncodeList([&labelList](const auto & encoder) -> CHIP_ERROR {
-            for (auto label : labelList)
+        err = aEncoder.EncodeList([&it](const auto & encoder) -> CHIP_ERROR {
+            CHIP_ERROR error = CHIP_NO_ERROR;
+            UserLabel::Structs::LabelStruct::Type userlabel;
+
+            while (it->Next(userlabel))
             {
-                ReturnErrorOnFailure(encoder.Encode(label));
+                SuccessOrExit(error = encoder.Encode(userlabel));
             }
 
-            return CHIP_NO_ERROR;
+        exit:
+            it->Release();
```

Comment:
This doesn't make any sense to me.  Don't we want to `it->Release()` any time `it` is not null, once we are done with it?  This is only releasing it on errors.

Reply:
Not sure I understand it, do not we always hit 'it->Release()'? if SuccessOrExit(error = encoder.Encode(userlabel)) check fail, we just jump to 'it->Release()' directly from the while loop.

Reply:
Just to be concrete: what happens if `it->Count() == 0`?

The logic should really be:
```
if (it) {
   // do stuff.
   it->Release();
} else {
  // whatever (probably error out?)
}
```